lib/commands/touch.js

```diff
@@ -0,0 +1,274 @@
+    case 'longPress':
+      if (typeof opts.duration === 'undefined' || !opts.duration) {
+        opts.duration = 1000;
+      }
+      return await this.touchLongClick(opts.element, opts.x, opts.y, opts.duration);
+    case 'cancel':
+      // TODO: clarify behavior of 'cancel' action and fix this
+      log.warn("Cancel action currently has no effect");
+      break;
+    default:
+      return `unknown action ${action}`;
+  }
+};
+
+
+// drag is *not* press-move-release, so we need to translate
+// drag works fine for scroll, as well
+commands.doTouchDrag = async function (gestures) {
+  let longPress = gestures[0];
+  let moveTo = gestures[1];
+  let startX = longPress.options.x || 0,
+      startY = longPress.options.y || 0,
+      endX = moveTo.options.x || 0,
+      endY = moveTo.options.y || 0;
+  if (longPress.options.element) {
+    let loc = await this.getLocationInView(longPress.options.element);
+    startX += loc.x || 0;
+    startY += loc.y || 0;
+  }
+  if (moveTo.options.element) {
+    let loc = await this.getLocationInView(moveTo.options.element);
+    endX += loc.x || 0;
+    endY += loc.y || 0;
+  }
+  let apiLevel =  await this.adb.getApiLevel();
+  // lollipop takes a little longer to get things rolling
+  let duration = apiLevel >= 5 ? 2 : 1;
+  // `drag` will take care of whether there is an element or not at that level
+  return await this.drag(startX, startY, endX, endY, duration, 1, longPress.options.element, moveTo.options.element);
+};
+
+// Fix last release action
+helpers.fixRelease = async function (gestures, release) {
+  //  let release = gestures[actions.length - 1];
+  // sometimes there are no options
+  release.options = release.options || {};
+  // nothing to do if release options are already set
+  if (release.options.element || (release.options.x && release.options.y)) {
+    return;
+  }
+  // without coordinates, `release` uses the center of the screen, which,
+  // generally speaking, is not what we want
+  // therefore: loop backwards and use the last command with an element and/or
+  // offset coordinates
+  let ref = _(gestures).chain().initial().filter(function (gesture) {
+    let opts = gesture.options;
+    return opts.element || (opts.x && opts.y);
+  }).last().value();
+  if (ref) {
+    let opts = ref.options || {};
+    if (opts.element) {
+      // we retrieve the element location, might be useful in
+      // case the element becomes invalid
+      let loc = await this.getLocationInView(opts.element);
+      let size = await this.getSize(opts.element);
+      release.options = {
+        element: opts.element,
+        x: loc.x + size.width / 2,
+        y: loc.y + size.height / 2
+      };
+    }
+    if (opts.x && opts.y) {
+      release.options = _.pick(opts, 'x', 'y');
+    }
+  }
+  return release;
+};
+
+// Perform one gesture
+helpers.performGesture = async function (gesture) {
+  let res = await this.doTouchAction(gesture.action, gesture.options || {});
```

Comment:
yes that is good catch. AndroidBootstrap does not send a JSONWP object, this code would not have worked :+1: copying code curse